pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java

```diff
@@ -144,4 +147,22 @@ protected void aggregate(TransformBlock transformBlock, int length, int function
   public AggregationGroupByResult getResult() {
     return new AggregationGroupByResult(_groupKeyGenerator, _aggregationFunctions, _groupByResultHolders);
   }
+
+  @Override
+  public Collection<TableResizer.fullIntermediateResult> trimGroupByResult(int threshold, TableResizer tableResizer) {
+    // Check if a trim is necessary
+    int keyNum = 0;
+    if (_hasMVGroupByExpression) {
+      keyNum = _mvGroupKeys.length;
+    } else {
+      keyNum = _svGroupKeys.length;
+    }
+    Iterator<GroupKeyGenerator.GroupKey> groupKeyIterator = _groupKeyGenerator.getGroupKeys();
+    if (keyNum > threshold && threshold != -1) {
+      return tableResizer.trimInSegmentResults(groupKeyIterator, _groupByResultHolders, threshold);
```

Comment:
threshold should be calculated based on the `limit` instead of hardcoded

Reply:
Now the threshold is calculated in the instance plan maker

> threshold should be calculated based on the `limit` instead of hardcoded